blockprod/src/detail/block_maker.rs

```diff
@@ -185,3 +185,460 @@ impl BlockMaker {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::tests::setup_blockprod_test;
+    use chainstate_types::{BlockIndex, BlockPreconnectData, ConsensusExtraData};
+    use crypto::random::make_pseudo_rng;
+    use mempool::{MempoolInterfaceMock, MempoolSubsystemInterface};
+
+    use chainstate::{
+        chainstate_interface::ChainstateInterface,
+        BlockError::{self, PrevBlockNotFound},
+        {ChainstateError, MockChainstateInterfaceMock},
+    };
+
+    use common::{
+        chain::{
+            block::{timestamp::BlockTimestamp, BlockReward, ConsensusData},
+            Block,
+        },
+        primitives::{BlockHeight, Id, H256},
+        time_getter::TimeGetter,
+        Uint256,
+    };
+
+    use super::*;
+
+    #[tokio::test]
+    async fn collect_transactions_subsystem_error() {
+        let (mut manager, chain_config, chainstate, _mempool) = setup_blockprod_test();
+
+        let mock_mempool = MempoolInterfaceMock::new();
+
+        let mock_mempool_subsystem = manager.add_subsystem_with_custom_eventloop("mock-mempool", {
+            let mock_mempool = mock_mempool.clone();
+            move |call, shutdn| async move {
+                mock_mempool.run(call, shutdn).await;
+            }
+        });
```

Comment:
I don't think you need a custom eventloop unless you're testing some special corner cases, for example mempool subsystem dying prematurely due to a panic.

Reply:
Ah ok. I was just following the way `node` adds the `mempool` subsystem. Ok cool, I'll fix these.

Reply:
For the tests where `MempoolInterfaceMock` were used, the call to the mocked `run()` was needed. Otherwise, the real `run()` was called instead.

Also, `mempool` seems to be a special case with `add_subsystem()` as it doesn't implement the `Subsystem` interface like the others i.e `shutdown()` doesn't exist, and so `add_subsystem_with_custom_eventloop()` was the only way I could see it added. If I'm wrong about this, let me know how and I'll be happy to fix.

Reply:
Is there anything preventing implementing `Subsystem` for `MempoolInterfaceMock`?

Reply:
I had a lot of issues when I tried this, but the main issue was that `manager.add_subsystem("mock", mock_mempool)` expects its second parameter to be:

    Box<dyn MempoolInterface>

This then returns:

    subsystem::Handle<Box<dyn MempoolInterface>>

However, now there is a type difference with the current `MempoolHandle`, as it's defined as:

    subsystem::Handle<dyn MempoolInterface>

Because of that type difference, I can't use the handle that `add_subsystem()` returns. Maybe I'm missing something simple, but I couldn't find an easy way to convert between types, and so creating a custom event look just like how it's done in in the Node runner was the simpler solution.

Reply:
Hmm okay it seems the best way of fixing this would be at subsystem or mempool level. Let's leave it like it is, at least for now.